File: Data Structure II/Day 9 - String/187. Repeated DNA Sequences.java
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        // Approach 1: using HashSet
        // Time O(n^2)
        // Space O(n)
        Set visited = new HashSet(), repeated = new HashSet();
        for (int i=0; i+9<s.length(); i++){
            String sub = s.substring(i, i+10);
            if (!visited.add(sub))
                repeated.add(sub);
        }
        return new ArrayList(repeated);

        // Apprach 2: rolling-hash
        // Time O(n)
        // Space O(n)
        Set<Integer> words = new HashSet<>();
        Set<Integer> doubleWords = new HashSet<>();
        List<String> rs = new ArrayList<>();
        char[] map = new char[26];
        map['A' - 'A'] = 0;
        map['C' - 'A'] = 1;
        map['G' - 'A'] = 2;
        map['T' - 'A'] = 3;
        for (int i=0; i<s.length()-9; i++){
            int v= 0;
            for (int j=i; j<i+10; j++){
                v<<=2;
                v+=map[s.charAt(j)-'A'];
            }
            if (!words.add(v) && doubleWords.add(v)){
                rs.add(s.substring(i, i+10));
            }
        }
        return rs;
    }
}

File: Data Structure II/Day 9 - String/187. Repeated DNA Sequences.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        # Approach 1: Hash Map Brute Force -> TLE for python T.T
        # Time O(n^2)
        # Space O(n)
        # ans = set()
        # for i in range(len(s)-9):
        #     ss = s[i:i+10]
        #     if ss not in ans:
        #         for j in range(i+1, len(s)-9):
        #             sx = s[j:j+10]
        #             if ss==sx:
        #                 ans.add(ss)
        #                 break
        # return list(ans)

        # Approach 2: Rolling Hash
        # Time O(n)
        # Space O(n)
        hashMap = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
        words = set()
        doubleWords = set()
        rs = []
        for i in range(len(s) - 9):
            v = 0
            for j in range(i, i + 10):
                v <<= 2
                v += hashMap[s[j]]
            if v in words:
                if v not in doubleWords:
                    doubleWords.add(v)
                    rs.append(s[i: i + 10])
            words.add(v)
        return rs